### DISCUSSION TITLE
Exception: Shape type 'Layout' not found

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [OrchardCMS/OrchardCore#10319] [Q&A]
I am trying to use OrchardCore withan ASP.NET Core project. I want to use the Theming, and Tenants features. To enabled all that I need I used configured my host app like this

```
services.AddOrchardCore()
        .WithTenants()
        .ConfigureServices((tenantServices, serviceProvider) =>
        {
            tenantServices.AddScoped<IShellFeaturesManager, ShellFeaturesManager>();
            tenantServices.AddScoped(c =>
            {
                // this will always return null since its not used!
                return default(IShellDescriptorFeaturesManager);
            });
            tenantServices.AddResourceManagement();
            tenantServices.AddTagHelpers<LinkTagHelper>();
            tenantServices.AddTagHelpers<MetaTagHelper>();
            tenantServices.AddTagHelpers<ResourcesTagHelper>();
            tenantServices.AddTagHelpers<ScriptTagHelper>();
            tenantServices.AddTagHelpers<StyleTagHelper>();
            tenantServices.AddScoped<OrchardCore.Data.Documents.IDocumentStore, DocumentStore>();
        })
        .AddMvc()
        .WithFeatures()
        .AddDocumentManagement()
        .AddTheming()
        .Configure((app, routes, serviceProvider) =>
        {
            app.UseAuthorization();
            app.UseAuthentication();
            routes.MapAreaControllerRoute("default", ModuleName, "{controller=Home}/{action=Index}/{id?}");
        });
```

I then created a Theme "almost identical to TheTheme in OC." I removed the Menu views since I am not using the CMS. I then injected the `ISiteThemeService` service into one of my controller, and I can verify that thethe current theme name is the name of my theme using the `GetCurrentThemeNameAsync()` method.

I removed the Views folder from my host project, but when I run the app I get the following error

```
Exception: Shape type 'Layout' not found
```

![image](https://user-images.githubusercontent.com/24724371/134269439-e786ba9f-1a86-48c3-b195-0d09f80793bd.png)

Here is the structure of the theme, as you can see the `Layout.cshtml` exists.

![image](https://user-images.githubusercontent.com/24724371/134269544-889fc2f7-f6c4-4062-b4c1-c25fcba1d210.png)

How can I correct this error?

**Comment by ns8482e** [comment]
You need to have your themes added in features  e.g. `.WithFeatures("Theme1", "Theme2")`.  Also add theme project's reference to main web project. If that still doesn't work then change `.WithFeatures("Theme1", "Theme2")` to `.AddTenantFeatures("Theme1", "Theme2")`

**Comment by hishamco** [comment]
I'm facing same issue, but when I reset the theme, then enable it and make it default everything works fine!! Any idea

**Reply by ns8482e** [reply]
Keep your  Project, Assembly and Nuget package naming same

**Reply by hishamco** [reply]
Could you please have a look to [this](https://github.com/OrchardCoreContrib/OrchardCoreContrib.Themes/tree/dev/src/OrchardCoreContrib.Themes.StylishPortfolio) if you have time

> Keep your Project, Assembly and Nuget package naming same

What if I planned to change the NuGet package name?

**Reply by ns8482e** [reply]
Try adding Id in manifest that match with your project name/assembly name/nuget package id

**Reply by hishamco** [reply]
I will leave NuGet for now, let me add the "Id" and check

Thanks

**Reply by hishamco** [reply]
> Try adding Id in manifest that match with your project name/assembly name/nuget package id

I though it will be set automatically. I set it with no luck ;(

**Reply by ns8482e** [reply]
Then change the name to match your project name/assembly

**Reply by hishamco** [reply]
It's already match, I did another theme without setting the Id, I'm not sure why this happen with Liquid?!!

**Reply by ns8482e** [reply]
I guess you need to change  following line 
```cs
    Name = "Stylish Portfolio",
```

With

```cs
    Name = "OrchardCoreContrib.Themes.StylishPortfolio",
```

**Reply by hishamco** [reply]
The "name" is used as a text to display in the themes page

**Reply by shawty** [reply]
I've just hit this same problem.  For me I created a theme project using the dotnet new command, then swapped out the "Layout.liquid" file for "Layout.cshtml" (I know razor better, so wanted to use that)

When I activated my theme, I got the exact same "missing shape" exception as mentioned above.

To fix it, I created a new "Layout.liquid" file containing some dummy content in the views folder, along side my "Layout.cshtml", then ran things, everything worked fine, and Orchard used my "Layout.cshtml" as I wanted it to, ignored the liquid file.

It seems that the key is having that liquid file... removing that, causes the exception.

**Reply by Skrypt** [reply]
🤔 It is not supposed to. I'm using a single Layout.cshtml file in one of my themes without any issue.
Could it be that the theme itself was not rebuilt and only rebuilt with hot reload?

**Reply by shawty** [reply]
@Skrypt nope, it was definitely a full rebuild, that was exactly my thought when it hit me, and it was only afterwards when I saw the exception, that I ended up here after searching for it in bing.

I even went so far as to drop right out of Rider, try it in VS, then switched back to rider, before also doing a full compile/build at the cmd prompt using dotnet too.

And yes I can confirm that single "Layout.cshtml" on it's own has previously worked, when I was mucking about learning things by copying themes from the master project source on GH, so I do know that it can be made to work that way.

Right now, this is what I have:

![image](https://user-images.githubusercontent.com/1876454/178538818-a8f9cd1d-b1e9-47a5-8d4c-a348c25dfca6.png)

If I remove that "Layout.liquid" file, it goes boom.

Also and I did find this wierd, but never really chased it up...  IF I change the theme to my custom theme, then stop & restart the app, when I return to admin, my custom theme is not the selected one, I have to go in and click "make current" again to apply it.  The built in themes however, remain set and stable when I do this.

No idea if that's related or not however, but it was a little odd.

**Reply by Skrypt** [reply]
Can I see your theme .csproj file?

**Reply by shawty** [reply]
Sure

![image](https://user-images.githubusercontent.com/1876454/178540381-a243423f-2ba8-4549-b7da-be8600c23846.png)

**Reply by Skrypt** [reply]
Ok, apart than the last ItemGroup should not be necessary everything seems fine.

**Reply by shawty** [reply]
ok...... this is real weird.  so, it appears I just managed to get it to work without the liquid file, but I'm not 100% sure what I did (Other than remove the liquid layout).

I did forcefully restart edge however, I mean proper force, task manager -> click, boom.... restart

and I also closed down skype, which was consuming about 2gb of ram!

Upon restarting edge, all seems to be ok now.

**Reply by shawty** [reply]
> 

When you say "The Last Item Group", I'm assuming you mean the item group holding the CSHTML files?

**Reply by Skrypt** [reply]
Also, to see the database records where it sets the themes you need to do : 

```SQL
SELECT * FROM Document WHERE [Type] = 'OrchardCore.Settings.SiteSettings, OrchardCore.Settings'
```

Then you can see what is happening with the Themes persisted in the database.
If what is set in the database is different than what you see then we may have an issue. Though if we can't repro clearly this could be a cache issue on your side.

**Reply by shawty** [reply]
AHHHH!! - It seems to be memory related.....

If I run some tasks and consume most of my memory, just enough so as windows doesn't start to go mental, but enough so as there's maybe only 1g or just under left, it starts playing up.

I'll see if I can come up with some concrete facts/figures, but 1g or less left seems to make edge + Orchard + Rider uncomfortable.

**Reply by shawty** [reply]
The retention of the theme setting seems to be an SQlite conflict.  I had the DB file open in Database Explorer for SQLite once I closed that, it saved ok.  That said, that was also closed AFTER I nuked edge, so it may have been partially due to the memory problem too.

I'll keep an eye on things and report back if/when I have any more.

**Reply by ns8482e** [reply]
@shawty   The shape `Layout.liquid`or `Layout.cshtml`  Needs to be embedded resource.  as @Skrypt suggested remove content from last item group . 

When you renamed and created cshtml it added it as content, that might work in dev but fail in prod

**Comment by FireEagle97** [comment]
Any new answer for this issue, I'm having the same issue when the file is called Layout.cshtml it works but when I add @Renderbody and move it to the Shared folder it gives me 'Layout' not found error @MikeAlhayek

**Reply by MikeAlhayek** [reply]
yes. if you are using MVC you 'll need to have it in a shared folder.

**Reply by FireEagle97** [reply]
I have it like that but it doesn't work
![err](https://github.com/OrchardCMS/OrchardCore/assets/62710297/c4918001-101e-4975-a63a-a4536a29f472)

**Reply by MikeAlhayek** [reply]
Does your _ViewStart have this?

Layout = "~/Views/Shared/_Layout.cshtml";

**Reply by FireEagle97** [reply]
no its
@{
 Layout = "_Layout"
}

**Reply by MikeAlhayek** [reply]
Try changing it and see if that works

**Reply by FireEagle97** [reply]
it didn't work

**Reply by FireEagle97** [reply]
@hishamco hi, I saw you had the same issue, can you help me solve it?

**Reply by hishamco** [reply]
https://github.com/OrchardCMS/OrchardCore/discussions/10319#discussioncomment-2083687

**Reply by FireEagle97** [reply]
@hishamco I think your case is different your theme uses liquid but mine uses MVC.

**Comment by t0mgerman** [comment]
Just in case anybody lands here from Google and this proves useful - I triggered this error by accidentally referencing my theme by two separate names in a startup recipe. The theme and a bunch of styled components are provided by a single module, which can be automatically enabled by choosing my recipe when creating a site. 

In the steps config, for the features list (the `enable` array) - I referred to my theme as `MyThemeName`. In the steps config, for the themes `site` setting (ie. the theme to use for the site itself) - I had referred to it as `OurCompanyName.MyThemeName`. It was the latter which was correct. So the fix was to ensure I was enabling the feature properly in steps. 

Guidance above helped me troubleshoot, I couldn't even query the DB when spinning up a site with a SQLite Instance because it didn't even get as far as creating the DB!

**Reply by Piedone** [reply]
Thanks for sharing!

**Reply by vip924** [reply]
We are getting the same error **"InvalidOperationException: The shape type 'Layout' is not found for the theme"** for our already existing tenant, when we diable the custom theme the error message theme name changes to the orchard core admin theme.

@t0mgerman any idea how to resolve

**Reply by vip924** [reply]
Adding this code to the program.cs solved the issue.

services.AddOrchardCore()
    .AddMvc()
    .WithFeatures("YourTheme");

**Comment by MikeKry** [comment]
This issue started occurring for us as well when using a custom CMS theme.

What’s strange is that it only happens on the default tenant. The theme works fine for a while, but after some time (likely after an application shutdown), it stops working.

The temporary fix is to recycle the application pool, but I haven’t been able to find a proper solution. It happens only in one project out of hundreds that use a custom theme — with the same structure and patterns. The only substantial difference is a name of the theme. Also, the issue occurs only in the development environment, not in production.

Did you ever find a solution @MikeAlhayek  @ns8482e  @hishamco  ?

**Reply by Piedone** [reply]
Since this error doesn't have a single cause, I don't think it's possible to tell what may be the issue there. However, I think there's a slight chance what you're seeing is the issue: https://github.com/OrchardCMS/OrchardCore/discussions/18342. This was fixed as part of 2.2.1: https://github.com/OrchardCMS/OrchardCore/pull/18370. Which version are you using?

**Reply by MikeAlhayek** [reply]
I haven’t looked into it since it’s not currently affecting us. However, in your project, check for any IShapeTableProvider instances registered within a feature in one of your modules. (So a module with multiple features and the provider is register in the feature not the module)

If an instance is registered under a feature (rather than directly in the module), make sure it’s decorated with the [Feature()] attribute. These providers are globally cached, and missing the attribute can lead to shape-related issues.

**Reply by hishamco** [reply]
I faced the same issue recently when I updated two of the OCC templates, I think it was an old bug caused by a shape. If you are still facing this issue let me know

**Reply by MikeKry** [reply]
It seems that updating to 2.2.1 (from 2.1.7) has solved the issue. Thank you so much to everyone :)
